qutebrowser/utils/log.py

```diff
@@ -24,16 +24,14 @@
 import contextlib
 import collections
 import copy
-import faulthandler
-import traceback
 import warnings
 import json
 import inspect
 import argparse
 from typing import (TYPE_CHECKING, Any, Iterator, Mapping, MutableSequence,
                     Optional, Set, Tuple, Union, TextIO, Literal, cast)
 
-from qutebrowser.qt import core as qtcore
+from qutebrowser.utils import qtlog
```

Comment:
As you've already mentioned, this basically just moves the problem, but doesn't fix it yet.

However, I like taking things step by step, so if you prefer we can also merge this as a first refactoring (once tests are green) and then take it from there in a second PR.

A quick look to what's next either way:

- Adding a test for the bug. `tests/end2end/test_invocations.py` is a good place for this, it has various tests for command-line arguments already. See e.g. `test_optimize` for some inspiration. I'm not sure off-hand how the testing machinery will break if too much is filtered out with `--logfilter`, might need to make sure that init/shutdown stuff is still logged since the testing code (in `quteprocess.py`) might look for those...
- Moving this to where `log.init_log()` is called in `qutebrowser/misc/earlyinit.py`. You might need to fix some tests in `tests/unit/utils/test_log.py`.
- Adding a comment here about being careful with imports, similarly to the `NOTE:` in `earlyinit.py`
- Possibly moving the `log.init_log()` up earlier in the init process

Reply:
I addressed most of your points. Although I'm still struggling a bit figuring out how to move `log.init_log()` up earlier in the init process.

To move `log.init_log()` *before* `machinery.init()`, we would have to extract [`log.init.debug(str(machinery.INFO))`](https://github.com/qutebrowser/qutebrowser/pull/7772/files#diff-a3f2f10f8adb634b3c7e2c6e18e468f4fea27185be7105f840d620783667a323R302) (e.g. move it to `earlyinit.early_init()` right after `machinery.init()`), otherwise we would keep crashing. But this requires importing `qutebrowser.utils.log` inside `early.early_init()`, and I am not sure if this is what we want.

Do you have some hints on how to proceed here?

Reply:
Could we just call `log.init.debug(str(machinery.INFO))` in `machinery.init()` (where it kinda belongs)? If importing `log.py` does not import any Qt stuff, doing that should be fine now.

Reply:
Thanks, that absolutely makes sense. I will try that tomorrow!